b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,384 @@
+  // Start TLS socket server.
+  // Return a promise for start() resolves/reject to
+  start: function() {
+    if (this._serverStatus == SERVER_STATUS.STARTED) {
+      return Promise.reject("AlreadyStarted");
+    }
+
+    let promise = new Promise((aResolve, aReject) => {
+      this._doStart(aResolve, aReject);
+    });
+    return promise;
+  },
+
+  // Stop TLS socket server, remove registered observer
+  // Cancel mDNS registration
+  // Return false if server not started, stop failed.
+  stop: function() {
+    if (this._serverStatus == SERVER_STATUS.STOPPED) {
+      return false;
+    }
+
+    if (!this._serverSocket) {
+      return false;
+    }
+
+    DEBUG && debug("Stop listening on port " + this._serverSocket.port);
+
+    Services.obs.removeObserver(this, "xpcom-shutdown");
+
+    if (this._mDNSRegistrationHandle) {
+      this._mDNSRegistrationHandle.Cancel(Cr.NS_OK);
+      this._mDNSRegistrationHandle = null;
+    }
+
+    this._port = -1;
+    this._serverSocket.close();
+    this._serverSocket = null;
+    this._serverStatus = SERVER_STATUS.STOPPED;
+
+    return true;
+  },
+
+  // Observers and Listeners
+  // nsIObserver
+  observe: function(subject, topic, data) {
+    switch (topic) {
+      case "xpcom-shutdown": {
+        // Stop service when xpcom-shutdown
+        this.stop();
+        break;
+      }
+    }
+  },
+
+  // nsIServerSocketListener
+  onSocketAccepted: function(socket, trans) {
+    DEBUG && debug("onSocketAccepted(socket=" + socket + ", trans=" + trans + ")");
+    DEBUG && debug("New connection on " + trans.host + ":" + trans.port);
+
+    const SEGMENT_SIZE = 8192;
+    const SEGMENT_COUNT = 1024;
+
+    try {
+      var input = trans.openInputStream(0, SEGMENT_SIZE, SEGMENT_COUNT)
+                       .QueryInterface(Ci.nsIAsyncInputStream);
+      var output = trans.openOutputStream(0, 0, 0);
+    } catch (e) {
+      DEBUG && debug("Error opening transport streams: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    let connectionId = ++nextConnectionId;
+
+    try {
+      // Create a connection for each user connection
+      // EventHandler implements nsIInputStreamCallback for incoming message from user
+      var conn = new Connection(input, output, this, connectionId);
+      let handler = new EventHandler(conn);
+
+      input.asyncWait(conn, 0, 0, Services.tm.mainThread);
+    } catch (e) {
+      DEBUG && debug("Error in initial connection: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    this._connections.set(connectionId, conn);
+    DEBUG && debug("Start connection " + connectionId);
+  },
+
+  // Close all connection when socket closed
+  onStopListening: function(socket, status) {
+    DEBUG && debug("Shut down server on port " + socket.port);
+
+    this._connections.forEach(function(connection){
+      connection.close();
+    });
+  },
+
+  // PRIVATE FUNCTIONS
+  _doStart: function(aResolve, aReject) {
```

Comment:
I saw mix of `aArg` and `arg` naming convention in this file, please choose one ( our coding style guide prefers `aArg`, see https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style#Methods)